src/Microsoft.VisualStudio.Editors/Common/ClientShimUtils.vb

```diff
@@ -0,0 +1,146 @@
+﻿' Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+Imports System.IO
+Imports Microsoft.VisualStudio.OLE.Interop
+Imports Microsoft.VisualStudio.ProjectSystem.Query
+Imports Microsoft.VisualStudio.ProjectSystem.Query.ProjectModel
+Imports Microsoft.VisualStudio.Shell.Interop
+Imports Microsoft.VisualStudio.ProjectSystem.Query.ProjectModel.Implementation
+Imports Microsoft.VisualStudio.ProjectSystem.Query.ServiceHub
+Imports Microsoft.VisualStudio.Shell
+
+Namespace Microsoft.VisualStudio.Editors.Common
+
+    ''' <summary>
+    ''' Shim layer to handle nexus and non-nexus APIs based on project
+    ''' </summary>
+    NotInheritable Class ClientShimUtils
+
+        Public Shared Function VSHierarchyShim(itemId As UInteger, VsHierarchy As IVsHierarchy, ServiceProvider As System.IServiceProvider) As IVsHierarchy
+           
+            'TODO - do we need to check if on nexus client ? If yes, how to check ?
```

Comment:
Today @tmeschter pointed me to `KnownUIContexts.CloudEnvironmentConnectedContext.IsActive` which may be helpful. I don't think it differentiates between client/server, but should differentiate between Codespaces/local.

Reply:
Thanks. It works.